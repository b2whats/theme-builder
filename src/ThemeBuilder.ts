import React from 'react'
import { css } from '@emotion/css'
import type { StyleProperties, StateSelectors } from './StyleProperties'
import { objectStyleToString } from './StyleProperties'
import { mergeObject } from './mergeTheme'
import { Tokens } from './tokens'

const Test = () => React.createElement('div')

type MaybeArray<Item> = Item extends []
  ? Item | Item[number]
  : Item[] | Item

export type Slot<Props = unknown> = unknown extends Props
  ? StyleProperties & StateSelectors & { withProps: boolean }
  : Props & { withProps: boolean, component: any }

type ShortHands = 'p' | 'px' | 'py' | 'm' | 'mx' | 'my'
export type ComponentTheme<Props, Slots> = {
  defaultProps?: Partial<Props>
  mapProps?: (props: Props) => Omit<Partial<Props>, ShortHands> | void | false
  slots: {
    [Name in keyof Slots]: {
      type?: 'tag' | 'component',
      styles: Parts<Props, Slots[Name]>[]
    }
  }
}

type ComputeTheme<Theme> = Theme extends ComponentTheme<infer _, infer Slots>
    ? {
      [Name in keyof Slots]: Slots[Name] extends Slot<infer Props>
        ? {
          type: 'component',
          props: Props
        } : {
          type: 'tag',
          className: string,
        } 
    }
    : never
type PropertyAsFunction<Map, Args> = { 
  [Key in keyof Map]?: NonNullable<Map[Key]> extends object
    ? PropertyAsFunction<Map[Key], Args>
    : Map[Key] | ((props: Args) => Map[Key])
}

type Parts<ExternalProps, SlotProps> = SlotProps extends StateSelectors
  ? string
    | PropertyAsFunction<SlotProps, ExternalProps>
    | ((props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string)
  : PropertyAsFunction<SlotProps, ExternalProps>

type SwitchMap<Props, Prop extends keyof Props, Result> = {
  [Value in Extract<Props[Prop], string>]?: Result | ((props: Props) => Result | undefined)
}

interface SlotUtils {
  if<ExternalProps, Props extends MaybeArray<keyof ExternalProps>, Values, PropertyValues extends Values>(
    prop: Props | Props[],
    value: PropertyValues,
    defaultValue?: PropertyValues,
  ): (props: ExternalProps) => Values | undefined
  map<ExternalProps, Prop extends keyof ExternalProps, Values, Default extends Values>(
    prop: Prop,
    body: SwitchMap<ExternalProps, Prop, Values>,
    defaultValue?: Default,
  ): (props: ExternalProps) => Values | undefined
  css<ExternalProps, SlotProps>(
    literals: TemplateStringsArray,
    ...placeholders: Array<(props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string>
  ): (props: ExternalProps, tokens: Tokens, slotStyles: SlotProps) => string
}

export const utils: SlotUtils = {
  if: (keys, value, defaultValue) => {
    return (props: any) => {
      const isValue = Array.isArray(keys)
        ? keys.every(key => Boolean(props[key]))
        : Boolean(props[keys])
      
      return isValue ? value : defaultValue
    }
  },
  map: (prop, options, defaultValue) => {
    return (props: any) => {
      const value = props[prop]

      return (typeof options[value] === 'function' ? options[value](props) : options[value]) || defaultValue
    }
  },
  css: (literals, ...placeholders) => {
    return function cssRewrite(props, tokens, slotStyles) {
      let result = ''

      for (let i = 0; i < placeholders.length; i++) {
          result += literals[i];
          result += placeholders[i](props, tokens, slotStyles)
      }
  
      result += literals[literals.length - 1];

      return result
    }
  }
}

const shortHands = {
  m: (value: any) => ({ mt: value, mr: value, mb: value, ml: value }),
  mx: (value: any) => ({ mr: value, ml: value }),
  my: (value: any) => ({ mt: value, mb: value }),
  p: (value: any) => ({ pt: value, pr: value, pb: value, pl: value }),
  px: (value: any) => ({ pr: value, pl: value }),
  py: (value: any) => ({ pt: value, pb: value }),
}

const isObject = (value: any): value is object => value && value.constructor && value.constructor === Object

function expandShorthands<T>(
  styles: T
): T {
  let results = {} as T

  if (!isObject(styles)) return styles

  for (let [key, value] of Object.entries(styles)) {
    if (shortHands[key]) {
      Object.assign(results, shortHands[key](value))

      continue
    }
    if (['disabled', 'hover', 'active', 'visited', 'checked'].includes(key)) {
      results[key] = expandShorthands(value)
    }

    results[key] = value
  }

  return results
}

const weight = (val: any): number => {
  if (typeof val === 'string') return 1
  if (typeof val === 'object') return 2
  if (typeof val === 'function') return 3

  return 0
}

export function mergeStyles<Styles extends Parts<any, any>>(arr: Styles[]): Styles[] {
  return arr.sort((a, b) => weight(a) - weight(b)).reduce<Styles[]>((acc, cur) => {
    const prev = acc[acc.length - 1]
    if (typeof prev !== typeof cur || typeof cur === 'function') {
      acc.push(cur)

      return acc
    }

    if (typeof prev === 'string' && typeof cur === 'string') {
      (acc[acc.length - 1] as string) = prev + cur

      return acc
    }
    if (typeof prev === 'object' && typeof cur === 'object') {
      (acc[acc.length - 1] as object) = mergeObject(prev as object, cur as object)

      return acc
    }

    return acc
  }, [])

}

export class ThemeBuilder<Theme extends ComponentTheme<any, any>> {
  theme = {
    slots: {}
  } as Theme

  cache = new WeakMap()

  slot<Name extends keyof this['theme']['slots']>(
    name: Name,
    styles: this['theme']['slots'][Name]['styles'] | this['theme']['slots'][Name]['styles'][number],
    type?: 'component' | 'tag'
  ) {
    this.theme.slots[name] = { 
      type,
      styles: Array.isArray(styles)
        ? mergeStyles((styles as []).map(expandShorthands))
        : [expandShorthands(styles)]
    }

    return this
  }
  defaultProps(props: this['theme']['defaultProps']) {
    if (!props) return this

    this.theme.defaultProps = expandShorthands(props)

    return this
  }
  mapProps(map: this['theme']['mapProps']) {
    this.theme.mapProps = map

    return this
  }
  merge(builder: this | undefined): this {
    if (!builder) return this

    if (this.cache.has(builder)) return this.cache.get(builder)

    const instance = new ThemeBuilder<Theme>()

    if (this.theme.defaultProps || builder.theme.defaultProps) {
      instance.defaultProps(
        Object.assign({}, this.theme.defaultProps, builder.theme.defaultProps)
      )
    }

    if (this.theme.mapProps || builder.theme.mapProps) {
      instance.mapProps((props) => ({
        ...this.theme.mapProps && this.theme.mapProps(props),
        ...builder.theme.mapProps && builder.theme.mapProps(props),
      }))
    }

    for (let [name, parts] of Object.entries(this.theme.slots)) {
      instance.slot(name, mergeStyles([
        ...parts.styles,
        ...builder.theme.slots[name] ? builder.theme.slots[name].styles : [],
      ]), parts.type)
    }

    this.cache.set(builder, instance)

    return instance as this
  }
  compute(props: this['theme']['defaultProps'], tokens: Tokens): ComputeTheme<this['theme']> {
    const result = {} as any

    props = {
      ...this.theme.defaultProps,
      ...expandShorthands(props),
    }

    if (this.theme.mapProps) {
      Object.assign(props, this.theme.mapProps(props))
    }
    
    for (let [name, { type, styles }] of Object.entries(this.theme.slots)) {
      if (type === 'tag') {
        let slotStyles = {}

        const className = styles.map(style => {
          if (typeof style === 'string') {
            return css(style)
          }
  
          if (typeof style === 'object') {
            if (style.withProps) {
              style = {
                ...style,
                ...props,
              }
            }
            slotStyles = style
            return css(objectStyleToString(props, tokens, style as Slot))
          }
  
          if (typeof style === 'function') {
            return css(style(props, tokens, slotStyles))
          }
  
          return ''
        }).join(' ')

        result[name] = {
          type,
          className,
        }
      } else {
        const componentProps = styles.map(style => {
          if (typeof style === 'object') {
            if (style.withProps) {
              style = {
                ...style,
                ...props,
              }

              delete style.withProps
            }

            const result = {}
            for (let prop in style) {
              result[prop] = typeof style[prop] === 'function'
                ? style[prop](props)
                : style[prop]
            }

            return result
          }

          return {}
        }).reduce((acc, curr) => {
          Object.assign(acc, curr)

          return acc
        })

        result[name] = {
          type,
          props: componentProps as any,
        }
      }

    }

    return result
  }
}