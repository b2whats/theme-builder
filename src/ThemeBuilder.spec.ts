import type { StyleProperties } from './StyleProperties'
import { ThemeBuilder, utils, mergeStyles, ComponentTheme, Slot } from './ThemeBuilder'

interface ButtonProps {
  children?: any
  iconBefore?: React.ReactNode
  iconAfter?: React.ReactNode
  size?: 's' | 'm' | 'l'
  kind?: 'fill' | 'outline' | 'flat',
  p?: number,
  mapProps1?: string,
  mapProps2?: string,
}

interface TextProps extends StyleProperties {
  fontSize?: number
  lineHeight?: number | 'none' | 'inherit'
  dense?: boolean
}

type ButtonTheme = ComponentTheme<ButtonProps, {
  Button: Slot,
  Text: Slot<TextProps>,
}>
  

test('defaultProps = undefined', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .defaultProps(undefined)

  expect(ButtonTheme).toMatchSnapshot()
})

test('defaultProps = { size: s }', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      size: 's'
    })

  expect(ButtonTheme).toMatchSnapshot()
})

test('mapProps = () => ({})', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .mapProps(() => ({}))

  expect(ButtonTheme).toMatchSnapshot()
})

test('slot Button is tag (default type)', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Button', {
      fontSize: 12
    }, 'div')

  expect(ButtonTheme).toMatchSnapshot()
})

test('slot Text is component', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Text', {
      fontSize: 12
    })

  expect(ButtonTheme).toMatchSnapshot()
})

test('slot Button as object styles', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Button', {
      fontSize: 12,
      color: () => 'red300',
    }, 'div')

  expect(ButtonTheme).toMatchSnapshot()
})

test('slot Button as array styles', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Button', [
      {
        fontSize: 12
      }
    ], 'div')

  expect(ButtonTheme).toMatchSnapshot()
})

test('slot Button as array styles with any elements', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Button', [
      '',
      {
        fontSize: 12
      },
      utils.css`
        color: red;
      `
    ], 'div')

  expect(ButtonTheme).toMatchSnapshot()
})

test('merge two theme - defaultProps empty/empty', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .defaultProps(undefined)

  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .defaultProps(undefined)

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)
  expect(ButtonTheme3).toMatchSnapshot()
})

test('merge two theme - defaultProps empty/not empty', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .defaultProps(undefined)

  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      size: 's'
    })

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)
  expect(ButtonTheme3).toMatchSnapshot()
})

test('merge two theme - defaultProps additional props', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      size: 's'
    })

  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      kind: 'flat'
    })

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)
  expect(ButtonTheme3).toMatchSnapshot()
})

test('merge two theme - defaultProps override props', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      size: 's'
    })

  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .defaultProps({
      size: 'l'
    })

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)
  expect(ButtonTheme3).toMatchSnapshot()
})

test('merge two theme - mapProps empty/not empty', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()

  const mapProps2 = jest.fn()
  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .mapProps(mapProps2)

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)

  const props: ButtonProps = {
    kind: 'flat'
  }
  ButtonTheme3.theme.mapProps!(props)

  expect(mapProps2).toHaveBeenCalledWith(props);
  expect(ButtonTheme3).toMatchSnapshot()
})

test('merge two theme - mapProps not empty/not empty', () => {
  const mapProps1 = jest.fn(({ kind }) => ({ mapProps1: kind }))
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .mapProps(mapProps1)
  const mapProps2 = jest.fn(({ kind }) => ({ kind, mapProps2: kind }))
  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .mapProps(mapProps2)

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)

  const props: ButtonProps = {
    kind: 'flat'
  }

  const mapProps3 = ButtonTheme3.theme.mapProps!(props)
  expect(mapProps1).toHaveBeenCalledWith(props);
  expect(mapProps2).toHaveBeenCalledWith(props);

  expect(mapProps3).toMatchSnapshot()
})

test('merge two theme - cache 2', () => {
  const ButtonTheme1 = new ThemeBuilder<ButtonTheme>()
    .slot('Button', {
      width: 100
    })
  const ButtonTheme2 = new ThemeBuilder<ButtonTheme>()
    .slot('Button', {
      width: 200
    })

  const ButtonTheme3 = ButtonTheme1.merge(ButtonTheme2)
  const ButtonTheme4 = ButtonTheme1.merge(ButtonTheme2)

  expect(ButtonTheme3).toBe(ButtonTheme4)
})

test('mergeStyles - sort', () => {
  const styles = mergeStyles(['', {}, () => {}, '', {}, {}, () => {}])

  expect(styles).toMatchSnapshot()
})

test('mergeStyles - merge', () => {
  const styles = mergeStyles([
    'str1;',
    { one: '1', two: 1 },
    () => {},
    'str2',
    { two: 2 },
    { three: '3', one: 3 },
    () => {}
  ])

  expect(styles).toMatchSnapshot()
})

test('mergeStyles - nested merge', () => {
  const styles = mergeStyles([
    'str1;',
    { one: '1', two: 1, nested: { one: '1', two: '1'} },
    () => {},
    'str2',
    { two: 2 },
    { three: '3', one: 3, nested: { three: 3, two: 3} },
  ])

  expect(styles).toMatchSnapshot()
})

test('compute', () => {
  const ButtonTheme = new ThemeBuilder<ButtonTheme>()
    .slot('Button', [
      {
        shrink: true,
        withProps: true,
      },
      {
        p: 12,
      },
    ], 'div')
    .slot('Text', [
      {
        withProps: true,
        grow: true,
      }
    ])

  const props: ButtonProps = {
    kind: 'flat',
    p: 10,
  }

  const a = ButtonTheme.compute({ breakpoint: [] } as any)(props)
  expect(a).toMatchSnapshot()
})