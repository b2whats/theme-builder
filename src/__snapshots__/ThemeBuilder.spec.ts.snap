// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compute 1`] = `
Object {
  "Button": Object {
    "className": "css-qhe590",
    "type": "tag",
  },
  "Text": Object {
    "props": Object {
      "grow": true,
      "kind": "flat",
      "pb": 10,
      "pl": 10,
      "pr": 10,
      "pt": 10,
    },
    "type": "component",
  },
}
`;

exports[`defaultProps = { size: s } 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "defaultProps": Object {
      "size": "s",
    },
    "slots": Object {},
  },
}
`;

exports[`defaultProps = undefined 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {},
  },
}
`;

exports[`mapProps = () => ({}) 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "mapProps": [Function],
    "slots": Object {},
  },
}
`;

exports[`merge two theme - cache 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Button": Object {
        "styles": Array [
          Object {
            "width": 200,
          },
        ],
        "type": undefined,
      },
    },
  },
}
`;

exports[`merge two theme - defaultProps additional props 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "defaultProps": Object {
      "kind": "flat",
      "size": "s",
    },
    "slots": Object {},
  },
}
`;

exports[`merge two theme - defaultProps empty/empty 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {},
  },
}
`;

exports[`merge two theme - defaultProps empty/not empty 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "defaultProps": Object {
      "size": "s",
    },
    "slots": Object {},
  },
}
`;

exports[`merge two theme - defaultProps override props 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "defaultProps": Object {
      "size": "l",
    },
    "slots": Object {},
  },
}
`;

exports[`merge two theme - mapProps empty/not empty 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "mapProps": [Function],
    "slots": Object {},
  },
}
`;

exports[`merge two theme - mapProps not empty/not empty 1`] = `
Object {
  "kind": "flat",
  "mapProps1": "flat",
  "mapProps2": "flat",
}
`;

exports[`mergeStyles - merge 1`] = `
Array [
  "str1;str2",
  Object {
    "one": 3,
    "three": "3",
    "two": 2,
  },
  [Function],
  [Function],
]
`;

exports[`mergeStyles - nested merge 1`] = `
Array [
  "str1;str2",
  Object {
    "nested": Object {
      "one": "1",
      "three": 3,
      "two": 3,
    },
    "one": 3,
    "three": "3",
    "two": 2,
  },
  [Function],
]
`;

exports[`mergeStyles - sort 1`] = `
Array [
  "",
  Object {},
  [Function],
  [Function],
]
`;

exports[`slot Button as array styles 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Button": Object {
        "styles": Array [
          Object {
            "fontSize": 12,
          },
        ],
        "type": "tag",
      },
    },
  },
}
`;

exports[`slot Button as array styles with any elements 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Button": Object {
        "styles": Array [
          "",
          Object {
            "fontSize": 12,
          },
          [Function],
        ],
        "type": "tag",
      },
    },
  },
}
`;

exports[`slot Button as object styles 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Button": Object {
        "styles": Array [
          Object {
            "color": [Function],
            "fontSize": 12,
          },
        ],
        "type": "tag",
      },
    },
  },
}
`;

exports[`slot Button is tag (default type) 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Button": Object {
        "styles": Array [
          Object {
            "fontSize": 12,
          },
        ],
        "type": "tag",
      },
    },
  },
}
`;

exports[`slot Text is component 1`] = `
ThemeBuilder {
  "cache": WeakMap {},
  "theme": Object {
    "slots": Object {
      "Text": Object {
        "styles": Array [
          Object {
            "fontSize": 12,
          },
        ],
        "type": "component",
      },
    },
  },
}
`;
